### PR TITLE
RHIDP-2172: Generic GitHub action and reusable workflow to be used to…

### PR DESCRIPTION
**Pull request by jesuino**
… export dynamic plugins from any repository

In this PR we separate the action from main so it can be called from a branch pull request. It also makes possible to configure the target repo, so it can be used with different repos that follow `backstage/community-plugins` repo structure.

The action to create pull requests to update plugins refs will also need to be updated since currently it is hardcoded to work only with `backstage/community-plugins` repo.

### CODE REVIEW

File: .github/workflows/export-dynamic.yaml [hunk 1]
```diff
@@ -1,31 +1,22 @@
...
-          - '20.x'
 
       janus-cli-version:
         description: Version of the janus-idp/cli package.
         type: string
-        required: false
+        required: true
         default: '^1.18.0'
 
-      overlay-branch:
```
Comment (davidfestal) [inline .github/workflows/export-dynamic.yaml:20]: Why did you remove this one completly ?

Comment (jesuino) [reply .github/workflows/export-dynamic.yaml:20]: I was keeping as default only the branch from where the job started, but I will modify and keep the logic to optionally setup an overlay branch.

**Comment by jesuino** [comment]
@davidfestal I brought back the parameter `overlay-repo`

File: .github/workflows/export-dynamic.yaml [hunk 2]
```diff
@@ -76,25 +72,15 @@ jobs:
...
-          else
-            NODE_VERSION=$(cat node-version 2> /dev/null)
-            NODE_VERSION=${NODE_VERSION:-20.x}
-            echo "NODE_VERSION=$NODE_VERSION" >> $GITHUB_OUTPUT
-            JANUS_CLI_VERSION=$(cat janus-cli-version 2> /dev/null)
-            JANUS_CLI_VERSION=${JANUS_CLI_VERSION:-1.18.0}
-            echo "JANUS_CLI_VERSION=$JANUS_CLI_VERSION" >> $GITHUB_OUTPUT
-            echo "UPLOAD_PROJECT_ON_ERROR='false'" >> $GITHUB_OUTPUT
-          fi
+          echo "NODE_VERSION=${{ github.event.inputs.node-version }}" >> $GITHUB_OUTPUT
```
Comment (davidfestal) [inline .github/workflows/export-dynamic.yaml:80]: Are these lines even useful ?
I guess we can directly use the `github.event.inputs.xxx` values later on.

File: .github/workflows/export-dynamic-caller.yaml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      - uses: actions/checkout@v4.2.2
+      - name: Set environment variables
+        id: set-env-vars
+        shell: bash
+        run: |
+          if [[ "${{  github.event.inputs.node-version }}" != "" ]]
+          then
+            echo "NODE_VERSION=${{ github.event.inputs.node-version }}" >> $GITHUB_OUTPUT
+          else
+            NODE_VERSION=$(cat node-version 2> /dev/null)
```
Comment (davidfestal) [inline .github/workflows/export-dynamic-caller.yaml:72]: I wonder whether we can even put this logic (reading the node-version from a file at the root of the overlay structure) in the reusable workflow, and thus make the `node-version` input of the reusable workflow optional, with an empty string as default value.
WDYT ?

File: .github/workflows/export-dynamic-caller.yaml [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            NODE_VERSION=${NODE_VERSION:-20.x}
+            echo "NODE_VERSION=$NODE_VERSION" >> $GITHUB_OUTPUT
+          fi
+          if [[ "${{  github.event.inputs.janus-cli-version }}" != "" ]]
+          then
+            echo "JANUS_CLI_VERSION=${{ github.event.inputs.janus-cli-version }}" >> $GITHUB_OUTPUT
+          else
+            JANUS_CLI_VERSION=$(cat janus-cli-version 2> /dev/null)
```
Comment (davidfestal) [inline .github/workflows/export-dynamic-caller.yaml:80]: Same as my comment above for the `node-version`

File: .github/workflows/export-dynamic-caller.yaml [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            JANUS_CLI_VERSION=${JANUS_CLI_VERSION:-1.13.1}
+            echo "JANUS_CLI_VERSION=$JANUS_CLI_VERSION" >> $GITHUB_OUTPUT
+          fi
+          if [[ "${{  github.event.inputs.plugins-repo }}" != "" ]]
+          then
+            echo "PLUGINS_REPO=${{ github.event.inputs.plugins-repo }}" >> $GITHUB_OUTPUT
+          else
+            PLUGINS_REPO=$(cat plugins-repo 2> /dev/null)
```
Comment (davidfestal) [inline .github/workflows/export-dynamic-caller.yaml:88]: Same as above

File: .github/workflows/export-dynamic-caller.yaml [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            echo "OVERLAY_BRANCH=${{ github.event.inputs.overlay-branch }}" >> $GITHUB_OUTPUT
+          else
+            OVERLAY_BRANCH=${OVERLAY_BRANCH:-''}
+            echo "OVERLAY_BRANCH=$OVERLAY_BRANCH" >> $GITHUB_OUTPUT
+          fi
+          if [[ "${{  github.event.inputs.upload-project-on-error }}" != "" ]]
+          then
+            echo "UPLOAD_PROJECT_ON_ERROR='${{ github.event.inputs.upload-project-on-error }}'" >> $GITHUB_OUTPUT
+          else
+            echo "UPLOAD_PROJECT_ON_ERROR='false'" >> $GITHUB_OUTPUT
```
Comment (davidfestal) [inline .github/workflows/export-dynamic-caller.yaml:103]: This can simply be managed by a default value on the reusable workflow I guess.

File: .github/workflows/export-dynamic-caller.yaml [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            PLUGINS_REPO=$(cat plugins-repo 2> /dev/null)
+            PLUGINS_REPO=${PLUGINS_REPO:-backstage/community-plugins}
+            echo "PLUGINS_REPO=$PLUGINS_REPO" >> $GITHUB_OUTPUT
+          fi
+          if [[ "${{  github.event.inputs.overlay-branch }}" != "" ]]
+          then
+            echo "OVERLAY_BRANCH=${{ github.event.inputs.overlay-branch }}" >> $GITHUB_OUTPUT
+          else
+            OVERLAY_BRANCH=${OVERLAY_BRANCH:-''}
+            echo "OVERLAY_BRANCH=$OVERLAY_BRANCH" >> $GITHUB_OUTPUT
```
Comment (davidfestal) [inline .github/workflows/export-dynamic-caller.yaml:97]: This can simply be managed by a default value on the reusable workflow I guess.

File: .github/workflows/export-dynamic-caller.yaml [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    outputs:
+      node-version: ${{ steps.set-env-vars.outputs.NODE_VERSION }}
+      janus-cli-version: ${{ steps.set-env-vars.outputs.JANUS_CLI_VERSION }}
+      plugins-repo: ${{ steps.set-env-vars.outputs.PLUGINS_REPO }}
+      workspaces: ${{ steps.gather-workspaces.outputs.workspaces }}
+      upload-project-on-error: ${{ steps.set-env-vars.outputs.UPLOAD_PROJECT_ON_ERROR }}
+      overlay-branch: ${{ steps.set-env-vars.outputs.OVERLAY_BRANCH }}
+    steps:
+      - uses: actions/checkout@v4.2.2
+      - name: Set environment variables
```
Comment (davidfestal) [inline .github/workflows/export-dynamic-caller.yaml:64]: According to the comments below, could it be that this whole step isn't necessary anymore ?

File: .github/workflows/export-dynamic-caller.yaml (hunk 5 above)
Comment (davidfestal) [inline .github/workflows/export-dynamic-caller.yaml:88]: Do we want to make this whole process multi-source-repo ?
It might finally be easily doable but completing the update-plugins-repo-refs to take in account the `plugins-repo` file at various steps.

But I guess we can keep it for a next PR

Comment (jesuino) [reply .github/workflows/export-dynamic-caller.yaml:88]: That's exactly the idea: keep it ready for the incoming changes to the `update-plugins-repo-refs` changes. This way we should not have to modify this later

**Comment by jesuino** [comment]
thanks for your review, @davidfestal 

I made the changes and it seems to work on my repo. Let me know if I should adjust something else. Thanks!

image

File: .github/workflows/export-dynamic.yaml [hunk 9]
```diff
@@ -1,31 +1,22 @@
...
-        description: Branch of the overlay structure (current branch by default).
-        type: string
-        required: false
         default: ''
-      
+
       upload-project-on-error:
         description: Upload the complete project as a workflow artifact in case of error in order to troubleshoot.
-        required: false
+        required: true
```
Comment (davidfestal) [inline .github/workflows/export-dynamic.yaml:19]: Is there a reason you changed this input to `required` ?
It seems strange to have a default value if the input is required.

File: .github/workflows/export-dynamic.yaml [hunk 10]
```diff
@@ -76,25 +71,35 @@ jobs:
...
             NODE_VERSION=${NODE_VERSION:-20.x}
             echo "NODE_VERSION=$NODE_VERSION" >> $GITHUB_OUTPUT
+          fi
+          if [[ "${{  github.event.inputs.janus-cli-version }}" != "" ]]
+          then
+            echo "JANUS_CLI_VERSION=${{ github.event.inputs.janus-cli-version }}" >> $GITHUB_OUTPUT
+          else
             JANUS_CLI_VERSION=$(cat janus-cli-version 2> /dev/null)
-            JANUS_CLI_VERSION=${JANUS_CLI_VERSION:-1.18.0}
+            JANUS_CLI_VERSION=${JANUS_CLI_VERSION:-1.13.1}
```
Comment (davidfestal) [inline .github/workflows/export-dynamic.yaml:92]: why `1.13.1` ?

File: .github/workflows/export-dynamic.yaml [hunk 11]
```diff
@@ -34,13 +25,17 @@ on:
...
-  pull_request:
-    branches:
-      - 'releases/**'
+      overlay-branch:
+        description: Branch of the overlay structure (current branch by default).
+        type: string
+        required: false
+        default: ''
 
 concurrency:
```
Comment (davidfestal) [inline .github/workflows/export-dynamic.yaml:40]: This `concurrency` block should probably be removed, since it's a `workflow-call` workflow

File: .github/workflows/export-dynamic.yaml [hunk 12]
```diff
@@ -54,19 +49,19 @@ jobs:
...
       janus-cli-version: ${{ steps.set-env-vars.outputs.JANUS_CLI_VERSION }}
+      plugins-repo: ${{ steps.set-env-vars.outputs.PLUGINS_REPO }}
       workspaces: ${{ steps.gather-workspaces.outputs.workspaces }}
-      upload-project-on-error: ${{ steps.set-env-vars.outputs.UPLOAD_PROJECT_ON_ERROR }}
       overlay-repo-ref: ${{ steps.set-overlay-repo-ref.outputs.OVERLAY_REPO_REF }}
 
     steps:
-      - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+      - uses: actions/checkout@v4.2.2
         if: ${{ github.event.inputs.overlay-branch == '' }}
```
Comment (davidfestal) [inline .github/workflows/export-dynamic.yaml:58]: Afair, the way to retrieve the inputs of a `workflow-call` workflow, is the following syntax:

`inputs.overlay-branch` (no `github.event.` prefix).

This stands for the whole file I guess.

**Comment by jesuino** [comment]
Hello @davidfestal thanks again for your review. I did update the files accordingly. Please let me know if you have any comment or if you see any issue.

**Review by davidfestal — APPROVED** [review]
